feat(exceptions): introduce custom exceptions for community cards and…

juliankuecken opened the pull request:
This pull request introduces significant updates to custom exceptions, removing all generic `IllegalStateException`s and `IllegalArgumentException`s and 'masking' them with custom names. The custom exceptions are all nested classes within the respective file/class they are used in.

### Issues
This PR
- resolves #60
- resolves #58

juliankuecken commented:
@dev-laky I did not dare to touch the one or two `IllegalStateExceptions` within the `PersistenceAdapter` or `/commands/` directory. I don't think this should be a problem in the future.

dev-laky commented:
> @dev-laky I did not dare to touch the one or two `IllegalStateExceptions` within the `PersistenceAdapter` or `/commands/` directory. I don't think this should be a problem in the future.



> @dev-laky I did not dare to touch the one or two `IllegalStateExceptions` within the `PersistenceAdapter` or `/commands/` directory. I don't think this should be a problem in the future.

I mean nothing should be broken as long as you update the tests and just change the exceptions. 

But will be fine if you don't have the time probably

dev-laky commented on src/main/kotlin/hwr/oop/projects/peakpoker/core/card/CommunityCards.kt:
```diff
@@ -1,15 +1,27 @@
 class CommunityCards(
   private val cards: MutableList<Card> = mutableListOf(),
 ) : Iterable<Card> by cards {
+  /**
+   * Exception thrown when duplicate cards are found in the community cards
+   */
+  class DuplicateCardException(message: String) : IllegalStateException(message)
+
+  /**
+   * Exception thrown when the number of community cards is invalid
+   */
+  class InvalidCardConfigurationException(message: String) : IllegalStateException(message)
+
+  /**
+   * Exception thrown when attempting to deal community cards in an invalid round phase
+   */
+  class InvalidDealPhaseException(message: String) : IllegalStateException(message)
```
why are the classes inside here???

dev-laky commented on src/main/kotlin/hwr/oop/projects/peakpoker/core/card/HoleCards.kt:
```diff
@@ -1,22 +1,30 @@
-    }
-    if (cards.distinct().size != cards.size) {
-      throw DuplicateCardException("Hole cards must not contain duplicates.")
+    /**
+     * Exception thrown when duplicate cards are found in the hole cards
+     */
+    class DuplicateCardException(message: String) : IllegalStateException(message)
+
+    /**
+     * Exception thrown when the number of hole cards is invalid
+     */
+    class InvalidCardConfigurationException(message: String) : IllegalStateException(message)
```
why are the classes inside???

dev-laky commented on src/main/kotlin/hwr/oop/projects/peakpoker/core/game/PokerRound.kt:
```diff
@@ -22,6 +17,40 @@ class PokerRound(
   private val bigBlindAmount: Int,
   private val smallBlindIndex: Int,
 ) : GameActionable {
+  /**
+   * Exception thrown when a player attempts a betting action without sufficient chips
+   */
+  class InsufficientChipsForBettingException(message: String) : IllegalStateException(message)
+
+  /**
+   * Thrown when an action is attempted on a player in an invalid state
+   */
+  class InvalidPlayerStateException(message: String) : IllegalStateException(message)
+
+  /**
+   * Exception thrown when a bet amount violates poker betting rules
+   */
+  class InvalidBetAmountException(message: String) : IllegalStateException(message)
+
+  /**
+   * Exception thrown when a player attempts an invalid call action
+   */
+  class InvalidCallActionException(message: String) : IllegalStateException(message)
+
+  /**
+   * Exception thrown when a player attempts an invalid check action
+   */
+  class InvalidCheckActionException(message: String) : IllegalStateException(message)
+
+  /**
+   * Exception thrown when a player with a specified name is not found in the game
+   */
+  class PlayerNotFoundException(message: String) : IllegalStateException(message)
+
+  /**
+   * Exception thrown when an invalid operation is attempted for the current round phase
+   */
+  class IllegalRoundPhaseException(message: String) : IllegalStateException(message)
```
great, but why inside ?

dev-laky commented on src/main/kotlin/hwr/oop/projects/peakpoker/core/hand/HandEvaluator.kt:
```diff
@@ -10,16 +10,32 @@ import hwr.oop.projects.peakpoker.core.card.HoleCards
 class HandEvaluator(
   private val communityCards: CommunityCards,
 ) {
+  /**
+   * Exception thrown when an empty player list is provided
+   */
+  class EmptyPlayerListEvaluationException(message: String) : IllegalArgumentException(message)
+
+  /**
+   * Exception thrown when an invalid number of cards is found
+   */
+  class InvalidCardCountException(message: String) : IllegalArgumentException(message)
+
+  /**
+   * Exception thrown when no valid hand can be found
+   */
+  class NoValidHandFoundException(message: String) : IllegalStateException(message)
+
```
tell my why, aint nothing but a heartache

dev-laky commented on src/main/kotlin/hwr/oop/projects/peakpoker/core/hand/PokerHand.kt:
```diff
@@ -9,22 +9,38 @@ import hwr.oop.projects.peakpoker.core.card.Card
  * @property cards The list of cards that make up the hand
  */
 class PokerHand(private val cards: List<Card>) : Iterable<Card> by cards {
+  /**
+   * Exception thrown when a poker hand does not contain exactly 5 cards
+   */
+  class InvalidHandSizeException(message: String) : Exception(message)
+
+  /**
+   * Exception thrown when a poker hand contains duplicate cards
+   */
+  class DuplicateCardException(message: String) : Exception(message)
```
in and out

juliankuecken commented:
This pull request refactors the exception handling in the `peakpoker` project by removing centralized exception classes and replacing them with localized exceptions specific to relevant classes. It also updates the code to use these new exceptions and improves the semantics of error handling across the codebase.

### Refactoring Exception Handling

#### Removal of Centralized Exception Classes:
* Removed all centralized exception classes (`GameException`, `BettingException`, `CardException`, `DeckException`, `GameStateException`, `PlayerException`) and their derived exceptions from the `core/exceptions` package. These have been replaced by localized exceptions within relevant classes. (`[[1]](diffhunk://#diff-21862f695ed96d0ae8237edb9006b447f74bab4abbcf17ecaadf68c14c47a6caL1-L3)`, `[[2]](diffhunk://#diff-67fec6fff8582a7adcce35d8e31b4d92bbb5cfb80e4cda63af56621eec4fb900L1-L26)`, `[[3]](diffhunk://#diff-de6e27125e84308390fd66ea4b4ed975eaaf00d6a5db3ab791bec090dc2a05cbL1-L17)`, `[[4]](diffhunk://#diff-1e1389ca8404578de2aaa85b44bc86fde796d2d31efecc64cf911297d6ef208cL1-L11)`, `[[5]](diffhunk://#diff-4c3f9744016e79252b80a96e82ae9a5b3776be56f79a31408736e3c999e0f700L1-L23)`, `[[6]](diffhunk://#diff-0970979b56ded9004817c9e2166182e6da7a25f1bb2e249cf31a867a08996b4eL1-L26)`)

#### Addition of Localized Exceptions:
* Added localized exceptions to `CommunityCards`, `HoleCards`, `PokerGame`, `PokerRound`, and `Deck` classes for handling errors specific to their functionality. These include exceptions like `DuplicateCardException`, `InvalidCardConfigurationException`, `InvalidDealPhaseException`, `GameEndedException`, `PlayerNotFoundException`, and others. (`[[1]](diffhunk://#diff-bec84287a889693200bebfbee96070c36d2e24e9fe7fef5cdb62e804c72f808aL4-R24)`, `[[2]](diffhunk://#diff-56955e61ffa9f383882daf1bf6afd88f6edbd87a99c4c3ab2b7c46d6492a0d0aR12-R21)`, `[[3]](diffhunk://#diff-605d3e6aae9949c409039d67a198b1a86e2bad0bea4f9b4440cc211c9961bde8R13-R44)`, `[[4]](diffhunk://#diff-47c245dc6aa90aca5c9aaf887d1b55146b688bcbfaced0acbbb4620a39aa2d3cR20-R53)`, `[[5]](diffhunk://#diff-2e1bb48e21abb43cd511a7b2169b29f53c69bc3a0abd1e0f83ab1cb1f2c03e63L6-R15)`)

### Code Updates for Exception Usage

#### PokerGame and PokerRound Updates:
* Updated `PokerGame` to use localized exceptions such as `NoActiveRoundException` and `GameEndedException` for game state errors. (`[[1]](diffhunk://#diff-605d3e6aae9949c409039d67a198b1a86e2bad0bea4f9b4440cc211c9961bde8R13-R44)`, `[[2]](diffhunk://#diff-605d3e6aae9949c409039d67a198b1a86e2bad0bea4f9b4440cc211c9961bde8L69-R90)`)
* Updated `PokerRound` to use localized exceptions like `InvalidCallActionException`, `InvalidCheckActionException`, and `InsufficientChipsForBettingException` for betting and round phase-related errors. (`[[1]](diffhunk://#diff-47c245dc6aa90aca5c9aaf887d1b55146b688bcbfaced0acbbb4620a39aa2d3cL143-R172)`, `[[2]](diffhunk://#diff-47c245dc6aa90aca5c9aaf887d1b55146b688bcbfaced0acbbb4620a39aa2d3cL208-R237)`, `[[3]](diffhunk://#diff-47c245dc6aa90aca5c9aaf887d1b55146b688bcbfaced0acbbb4620a39aa2d3cL253-R288)`, `[[4]](diffhunk://#diff-47c245dc6aa90aca5c9aaf887d1b55146b688bcbfaced0acbbb4620a39aa2d3cL274-R309)`, `[[5]](diffhunk://#diff-47c245dc6aa90aca5c9aaf887d1b55146b688bcbfaced0acbbb4620a39aa2d3cL295-R324)`, `[[6]](diffhunk://#diff-47c245dc6aa90aca5c9aaf887d1b55146b688bcbfaced0acbbb4620a39aa2d3cL319-R350)`, `[[7]](diffhunk://#diff-47c245dc6aa90aca5c9aaf887d1b55146b688bcbfaced0acbbb4620a39aa2d3cL345-R374)`)

#### CommunityCards and HoleCards Updates:
* Replaced generic `IllegalStateException` with localized exceptions (`InvalidDealPhaseException`, `DuplicateCardException`, etc.) for better error specificity in the `CommunityCards` and `HoleCards` classes. (`[[1]](diffhunk://#diff-bec84287a889693200bebfbee96070c36d2e24e9fe7fef5cdb62e804c72f808aL25-R37)`, `[[2]](diffhunk://#diff-bec84287a889693200bebfbee96070c36d2e24e9fe7fef5cdb62e804c72f808aL38-R50)`, `[[3]](diffhunk://#diff-56955e61ffa9f383882daf1bf6afd88f6edbd87a99c4c3ab2b7c46d6492a0d0aR12-R21)`)

#### Deck Updates:
* Added `InsufficientCardsException` as a localized exception in the `Deck` class to handle errors when drawing more cards than available. (`[src/main/kotlin/hwr/oop/projects/peakpoker/core/deck/Deck.ktL6-R15](diffhunk://#diff-2e1bb48e21abb43cd511a7b2169b29f53c69bc3a0abd1e0f83ab1cb1f2c03e63L6-R15)`)

#### Command Updates:
* Updated `CreateNewGame` command to replace `GameException` with `IllegalStateException` for error handling during game creation. (`[[1]](diffhunk://#diff-08d516885ecbddc52c174a6fa59bb3479a1531a4478206db40d3420b215e17d0L9)`, `[[2]](diffhunk://#diff-08d516885ecbddc52c174a6fa59bb3479a1531a4478206db40d3420b215e17d0L34-R33)`)

juliankuecken commented:
@dev-laky Good morning sir, i had now a real deep dive into the tests once more (it's late at night by now) and i really must say: Congratulations! 👏🎉

dev-laky commented on src/main/kotlin/hwr/oop/projects/peakpoker/core/game/PokerRound.kt:
```diff
@@ -140,7 +169,7 @@ class PokerRound(
       RoundPhase.TURN -> RoundPhase.RIVER
       RoundPhase.RIVER -> RoundPhase.SHOWDOWN
-      RoundPhase.SHOWDOWN -> throw IllegalStateException("PokerGame is already in the SHOWDOWN phase")
+      RoundPhase.SHOWDOWN -> throw IllegalRoundPhaseException("PokerGame is already in the SHOWDOWN phase")
```
sorryyy

dev-laky reviewed (approved):
lgtm
